Cython/TestUtils.py

```diff
@@ -228,21 +229,42 @@ def unpack_source_tree(tree_file, workdir, cython_root):
     return workdir, header
 
 
+def fresh_cythonize(*args, **kwargs):
+    """Clear function caches and run cythonize"""
+    clear_function_and_Dependencies_caches()
+    Dependencies.cythonize(*args, **kwargs)
+
+
+def clear_function_and_Dependencies_caches():
+    Utils.clear_function_caches()
+    Dependencies._dep_tree = None  # discard method caches
+
+
 def write_file(file_path, content, dedent=False):
     """
-    Write some content (text or bytes) to the file at "file_path".
+    Write some content (text or bytes) to the file at `file_path` in utf-8.
     """
-    mode = 'wb' if isinstance(content, bytes) else 'w'
+
+    kwargs = {}
+
+    mode = "w"
+    if isinstance(content, bytes):
+        mode += "b"
+    else:
```

Comment:
I think we can afford spelling this out.
```suggestion
    if isinstance(content, bytes):
        mode = "wb"
    else:
        mode = "w"
```

Reply:
What if we want to support more modes, i.e.  not only `w`, but also `a`, for example?  Add conditions for every string / byte and `w` / `a` use case?

one could do it like this:
```
    mode = "w"
    if <use a>:
        mode = "a"

    if isinstance(content, bytes):
        mode += "b"
    else:
```

Reply:
Of course, I am theorizing again, but still.

Reply:
> What if we want to support more modes

Then, at the point where someone writes a test that needs this functionality, that person can change the function in the necessary way. Code is not carved in stone once it's written. Even Voyager-1 received software updates along the way.

Let's keep things as simple as possible, until we know that we need to accommodate for a change that requires more complexity. YAGNI, unless proven otherwise.

(I am aware that I am myself guilty of adding the "dedent" parameter without a strict need. I was only guessing that files with more than one line of code were likely to come up in the tests. Maybe that was a mistake. Maybe I should not have done that. You can probably tell better than I can at the moment.)

Reply:
> Even Voyager-1 received software updates along the way.

Well it seems that when discussing how to make a wrapper function to write to files and someone mentions Voyager ... well it seems the discussion has gone a little too far ;)

> YAGNI, unless proven otherwise.

> (I am aware that I am myself guilty of adding the "dedent" parameter without a strict need

Actually, if you look at it from this point of view, then there is another YAGNI - the binary mode itself... yeah

I just think that if you can predict some future changes and make them easier, or help a future programmer figure it out one way or another, then that's good.

> that requires more complexity.

IMHO my solution does not add any particular complexity

For me right now, it looks more like an opinion-based thingie. You think one implementation is better and easier to understand, but I think the other ...
Well, in this situation, I don't care what solution the end result is, and if you want, you can push yours. So yeah.

Reply:
So should we go back to basics and make a simple solution without functionality that nobody asked for?

Reply:
YAGNI is very slippery ice ...

Reply:
> there is another YAGNI - the binary mode itself

I left that out as a topic in my comment above because I though that one example would be enough.

I get the general feeling that we keep wasting time on discussing things in this PR that we don't need. We should try to keep our work more focussed, both yours and mine.

Reply:
> We should try to keep our work more focussed, both yours and mine.

Let's try